mailpoet/lib/Mailer/MailerLog.php

```diff
@@ -129,54 +196,128 @@ public static function setError($mailerLog, $operation, $errorMessage, $errorCod
   }
 
-  public static function incrementSentCount() {
+  /**
+   * @return MailerLogData|null
+   */
+  public static function incrementSentCount(): ?array {
     $mailerLog = self::getMailerLog();
     // do not increment count if sending limit is reached
-    if (self::isSendingLimitReached($mailerLog)) return;
+    if (self::isSendingLimitReached($mailerLog)) {
+      return null;
+    }
     // clear previous retry count, errors, etc.
-    if ($mailerLog['error']) {
+    if ($mailerLog['error'] !== null) {
       $mailerLog = self::clearSendingErrorLog($mailerLog);
     }
-    (int)$mailerLog['sent']++;
+
+    $time = self::sentEntriesDate();
+    if (!isset($mailerLog['sent'][$time])) {
+      $mailerLog['sent'][$time] = 0;
+    }
+    $mailerLog['sent'][$time]++;
     return self::updateMailerLog($mailerLog);
   }
 
-  public static function clearSendingErrorLog($mailerLog) {
+  /**
+   * @param MailerLogData $mailerLog
+   * @return MailerLogData
+   */
+  public static function clearSendingErrorLog(array $mailerLog): array {
     $mailerLog['retry_attempt'] = null;
     $mailerLog['retry_at'] = null;
     $mailerLog['error'] = null;
     return self::updateMailerLog($mailerLog);
   }
 
-  public static function isSendingLimitReached($mailerLog = false) {
+  /**
+   * @param MailerLogData|null $mailerLog
+   * @return bool
+   */
+  public static function isSendingLimitReached(array $mailerLog = null): bool {
     $settings = SettingsController::getInstance();
     $mailerConfig = $settings->get(Mailer::MAILER_CONFIG_SETTING_NAME);
     // do not enforce sending limit for MailPoet's sending method
     if ($mailerConfig['method'] === Mailer::METHOD_MAILPOET) return false;
     $mailerLog = self::getMailerLog($mailerLog);
-    $elapsedTime = time() - (int)$mailerLog['started'];
 
-    if (empty($mailer['frequency'])) {
+    if (empty($mailerConfig['frequency'])) {
       $defaultSettings = $settings->getAllDefaults();
-      $mailer['frequency'] = $defaultSettings['mta']['frequency'];
+      $mailerConfig['frequency'] = $defaultSettings['mta']['frequency'];
     }
     $frequencyInterval = (int)$mailerConfig['frequency']['interval'] * Mailer::SENDING_LIMIT_INTERVAL_MULTIPLIER;
     $frequencyLimit = (int)$mailerConfig['frequency']['emails'];
+    $sent = self::sentSince($frequencyInterval, $mailerLog);
+    return $sent >= $frequencyLimit;
+  }
 
-    if ($mailerLog['sent'] >= $frequencyLimit) {
-      if ($elapsedTime <= $frequencyInterval) return true;
-      // reset mailer log as enough time has passed since the limit was reached
-      self::resetMailerLog();
-    }
-    return false;
+  /**
+   * @param int $sinceSeconds
+   * @param MailerLogData|null $mailerLog
+   * @return int
+   */
+  private static function sentSince(int $sinceSeconds, array $mailerLog = null): int {
+
+    $sinceDate = date('Y-m-d H:i:s', time() - $sinceSeconds);
+    $mailerLog = self::getMailerLog($mailerLog);
+
+    return (int)array_sum(
```

Comment:
Hi :wave:,

Just visiting this repo and apparently subscribed to all PRs... But have you considered a "leaky bucket" algorithm which only requires you to keep track of a single int and the last time you put something in the bucket (sent an email)?

Reply:
Hi @withinboredom,
thanks for chiming in. I am not sure yet about your suggestion. My first impression is, if we take the last time we sent an email, we do not know, when this timespan started. So if we say we want to send 100/5 minutes, once we reached 100, we would need to wait 5 minutes before we can start again, so we could only send 100/10 minutes. But probably I miss the point. Can you elaborate more?

Reply:
Hi, @withinboredom! Did you by any chance mean to say "[token bucket](https://en.wikipedia.org/wiki/Token_bucket)" instead of "leaky bucket"?

Leaky bucket leaks at a constant rate and needs a buffer (FIFO queue) for that. For us, this would mean always sending at a constant rate, which is not what we do or want to do.

The token bucket algorithm adds tokens at a constant rate and can use them variably. It will spread out the sending over the period uniformly, which is again not something we need or want to do.

I guess we need a throttling based on a "sliding window" algorithm, which is what this PR implements. For the sliding window, the best is to have an exact timestamp of every sent email, which we currently don't have (some transactional emails) so we need to store the counts. We can take it as a temporary solution later to be replaced by full sent email log.